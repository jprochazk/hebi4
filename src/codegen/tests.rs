use std::fs::read_to_string;

use super::emit;
use crate::{parser::parse, token::tokenize};

uitest::uitest!("tests/inputs/codegen/*.hi", "codegen", |path| {
    let input = read_to_string(path).unwrap();
    let tokens = tokenize(&input);
    let ast = match parse(&tokens) {
        Ok(ast) => ast,
        Err(err) => {
            panic!("{}", err.render(&input));
        }
    };
    let (snapshot, failure) = match emit(&ast) {
        Ok(chunk) => (format!("OK\n{}", chunk.disasm(&input)), false),
        Err(err) => (format!("ERROR\n{}", err.render(&input)), true),
    };

    #[cfg(not(miri))]
    {
        insta::assert_snapshot!(snapshot);
        let _ = failure;
    }

    #[cfg(miri)]
    {
        if failure {
            panic!("{snapshot}");
        } else {
            eprintln!("{snapshot}");
        }
    }
});
