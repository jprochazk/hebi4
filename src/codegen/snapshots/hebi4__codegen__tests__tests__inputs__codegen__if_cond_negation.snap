---
source: src/codegen/tests.rs
expression: snapshot
---
SOURCE
let a = 0
let b = 1

if not (a < b) {}   // isge
if not (a > b) {}   // isle
if not (a <= b) {}  // isgt
if not (a >= b) {}  // islt
if not (a == b) {}  // isne
if not (a != b) {}  // iseq


OK
.fn @main
  params 0
  stack 4
  .locals
    a = r1
    b = r2
  .code
   0  lsmi r1, 0
   1  lsmi r2, 1
   2  isge r1, r2
   3  jmp 2   ; to 5
   4  lnil r3
   5  isle r1, r2
   6  jmp 2   ; to 8
   7  lnil r3
   8  isgt r1, r2
   9  jmp 2   ; to 11
  10  lnil r3
  11  islt r1, r2
  12  jmp 2   ; to 14
  13  lnil r3
  14  isne r1, r2
  15  jmp 2   ; to 17
  16  lnil r3
  17  iseq r1, r2
  18  jmp 2   ; to 20
  19  lnil r0
  20  ret
